Flank missing HTTP header X-Goog-User-Project

**Describe the bug**
> A clear and concise description of what the bug is.

Flank is failing with http error code 403 when using end user credentials to authenticate and run tests. Based on the error the HTTP requests are missing the 'X-Goog-User-Project' header which are required for quota and billing purposes.

<br/>

**To Reproduce**
> Steps to reproduce the behavior:

1. Run the command in terminal
   > gcloud auth application-default login
2. Run the commands within the android project and follow the instructions
   > ./gradlew flankAuth
   > ./gradlew runFlank
4. Error Returned:
`  "message": "Your application has authenticated using end user credentials from the Google Cloud SDK or Google Cloud Shell which are not supported by the testing.googleapis.com. We recommend configuring the billing/quota_project setting in gcloud or using a service account through the auth/impersonate_service_account setting. For more information about service accounts and how to use them in your application, see https://cloud.google.com/docs/authentication/. If you are getting this error with curl or similar tools, you may need to specify 'X-Goog-User-Project' HTTP header for quota and billing purposes. For more information regarding 'X-Goog-User-Project' header, please check https://cloud.google.com/apis/docs/system-parameters.",
  "status": "PERMISSION_DENIED"`

<br/>

**Expected behavior**
> A clear and concise description of what you expected to happen.

Flank should pass the required header (X-Goog-User-Project) irrespective of the login type (i.e. End User login or Service Account) and users must be able to run tests.

<br/>

**Details (please complete the following information):**
> Have you tested on the latest Flank snapshot?
Yes

> Post the output of ` flank --version`.
version: v23.01.0
revision: 7ad3d5ba4097f189965110ca52b93fd0645715e0


<br/>

**Additional context**
> Add any other context about the problem here.

Complete Error Log:

```
Mar 08, 2023 3:04:01 AM com.google.auth.oauth2.DefaultCredentialsProvider warnAboutProblematicCredentials
WARNING: Your application has authenticated using end user credentials from Google Cloud SDK. We recommend that most server applications use service accounts instead. If your application continues to use end user credentials from Cloud SDK, you might receive a "quota exceeded" or "API not enabled" error. For more information about service accounts, see https://cloud.google.com/docs/authentication/.
ftl.run.exception.PermissionDenied: Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 403 Forbidden
GET https://testing.googleapis.com/v1/testEnvironmentCatalog/ios?projectId=abc-12345678
{
  "code": 403,
  "details": [
    {
      "@type": "type.googleapis.com/google.rpc.ErrorInfo",
      "reason": "SERVICE_DISABLED"
    }
  ],
  "errors": [
    {
      "domain": "usageLimits",
      "message": "Your application has authenticated using end user credentials from the Google Cloud SDK or Google Cloud Shell which are not supported by the testing.googleapis.com. We recommend configuring the billing/quota_project setting in gcloud or using a service account through the auth/impersonate_service_account setting. For more information about service accounts and how to use them in your application, see https://cloud.google.com/docs/authentication/. If you are getting this error with curl or similar tools, you may need to specify 'X-Goog-User-Project' HTTP header for quota and billing purposes. For more information regarding 'X-Goog-User-Project' header, please check https://cloud.google.com/apis/docs/system-parameters.",
      "reason": "accessNotConfigured",
      "extendedHelp": "https://console.developers.google.com"
    }
  ],
  "message": "Your application has authenticated using end user credentials from the Google Cloud SDK or Google Cloud Shell which are not supported by the testing.googleapis.com. We recommend configuring the billing/quota_project setting in gcloud or using a service account through the auth/impersonate_service_account setting. For more information about service accounts and how to use them in your application, see https://cloud.google.com/docs/authentication/. If you are getting this error with curl or similar tools, you may need to specify 'X-Goog-User-Project' HTTP header for quota and billing purposes. For more information regarding 'X-Goog-User-Project' header, please check https://cloud.google.com/apis/docs/system-parameters.",
  "status": "PERMISSION_DENIED"
}
        at ftl.http.ExecuteWithRetryKt$executeWithRetry$$inlined$withRetry$1.invokeSuspend(ExecuteWithRetry.kt:30)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
        at ftl.http.ExecuteWithRetryKt.executeWithRetry(ExecuteWithRetry.kt:44)
        at ftl.client.google.IosCatalog.iosDeviceCatalog(IosCatalog.kt:53)
        at ftl.client.google.IosCatalog.getModels(IosCatalog.kt:21)
        at ftl.client.google.AndroidCatalog.isVirtualDevice(AndroidCatalog.kt:51)
        at ftl.args.PrepareAndroidCommonConfigKt.resolveDeviceType(PrepareAndroidCommonConfig.kt:14)
        at ftl.args.PrepareAndroidCommonConfigKt.prepareAndroidCommonConfig(PrepareAndroidCommonConfig.kt:10)
        at ftl.args.CreateAndroidArgsKt.createAndroidArgs$default(CreateAndroidArgs.kt:15)
        at ftl.domain.RunTestAndroidKt.invoke(RunTestAndroid.kt:41)
        at ftl.presentation.cli.firebase.test.android.AndroidRunCommand.run(AndroidRunCommand.kt:58)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
        at picocli.CommandLine.access$1300(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        at ftl.Main$main$1.invoke(Main.kt:12)
        at ftl.Main$main$1.invoke(Main.kt:10)
        at ftl.run.exception.ExceptionHandlerKt.withGlobalExceptionHandling(ExceptionHandler.kt:24)
        at ftl.run.exception.ExceptionHandlerKt.withGlobalExceptionHandling(ExceptionHandler.kt:14)
        at ftl.Main.main(Main.kt:10)

Total run duration: 0m  1s
```

--- a/test_runner/src/main/kotlin/ftl/client/google/AndroidCatalog.kt
+++ b/test_runner/src/main/kotlin/ftl/client/google/AndroidCatalog.kt
@@ -1,9 +1,11 @@
 package ftl.client.google
 
+import com.google.api.client.http.HttpHeaders
 import com.google.testing.model.AndroidDevice
 import com.google.testing.model.AndroidDeviceCatalog
 import com.google.testing.model.AndroidModel
 import com.google.testing.model.Orientation
+import ftl.config.FtlConstants.GCS_PROJECT_HEADER
 import ftl.http.executeWithRetry
 import ftl.presentation.cli.firebase.test.reportmanager.ReportManagerState
 import ftl.presentation.publish
@@ -21,6 +23,7 @@ object AndroidCatalog {
         GcTesting.get.testEnvironmentCatalog()
             .get("android")
             .setProjectId(projectId)
+            .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, projectId))
             .executeWithRetry()
             .androidDeviceCatalog
             .filterDevicesWithoutSupportedVersions()

--- a/test_runner/src/main/kotlin/ftl/client/google/GcTestMatrix.kt
+++ b/test_runner/src/main/kotlin/ftl/client/google/GcTestMatrix.kt
@@ -1,7 +1,9 @@
 package ftl.client.google
 
+import com.google.api.client.http.HttpHeaders
 import com.google.testing.model.CancelTestMatrixResponse
 import com.google.testing.model.TestMatrix
+import ftl.config.FtlConstants.GCS_PROJECT_HEADER
 import ftl.http.executeWithRetry
 import ftl.run.exception.FlankGeneralError
 import kotlinx.coroutines.Dispatchers
@@ -38,6 +40,7 @@ object GcTestMatrix {
     suspend fun refresh(testMatrixId: String, projectId: String): TestMatrix {
         val getMatrix = withContext(Dispatchers.IO) {
             GcTesting.get.projects().testMatrices().get(projectId, testMatrixId)
+                .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, projectId))
         }
         var failed = 0
         val maxWait = ofHours(1).seconds
@@ -57,6 +60,7 @@ object GcTestMatrix {
     suspend fun cancel(testMatrixId: String, projectId: String): CancelTestMatrixResponse {
         val cancelMatrix = withContext(Dispatchers.IO) {
             GcTesting.get.projects().testMatrices().cancel(projectId, testMatrixId)
+                .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, projectId))
         }
         var failed = 0
         val maxTries = 3

--- a/test_runner/src/main/kotlin/ftl/client/google/GcToolResults.kt
+++ b/test_runner/src/main/kotlin/ftl/client/google/GcToolResults.kt
@@ -1,5 +1,6 @@
 package ftl.client.google
 
+import com.google.api.client.http.HttpHeaders
 import com.google.api.services.toolresults.ToolResults
 import com.google.api.services.toolresults.model.Environment
 import com.google.api.services.toolresults.model.Execution
@@ -16,6 +17,7 @@ import com.google.testing.model.ToolResultsHistory
 import com.google.testing.model.ToolResultsStep
 import ftl.args.IArgs
 import ftl.config.FtlConstants
+import ftl.config.FtlConstants.GCS_PROJECT_HEADER
 import ftl.config.FtlConstants.JSON_FACTORY
 import ftl.config.FtlConstants.applicationName
 import ftl.config.FtlConstants.httpTransport
@@ -44,6 +46,7 @@ object GcToolResults {
             .histories()
             .list(args.project)
             .setFilterByName(args.resultsHistoryName)
+            .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, args.project))
             .executeWithRetry()
         return result?.histories ?: emptyList()
     }
@@ -56,6 +59,7 @@ object GcToolResults {
             .projects()
             .histories()
             .create(args.project, history)
+            .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, args.project))
             .execute()
     }
 
@@ -90,6 +94,7 @@ object GcToolResults {
                 toolResultsStep.historyId,
                 toolResultsStep.executionId
             )
+            .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, testExecution.projectId))
             .executeWithRetry()
     }
 
@@ -105,6 +110,7 @@ object GcToolResults {
                 toolResultsStep.executionId,
                 toolResultsStep.stepId
             )
+            .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, toolResultsStep.projectId))
             .executeWithRetry()
     }
 
@@ -121,6 +127,7 @@ object GcToolResults {
             toolResultsStep.executionId,
             toolResultsStep.stepId
         )
+        .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, toolResultsStep.projectId))
         .executeWithRetry()
 
     // Lists Test Cases attached to a Step
@@ -141,6 +148,7 @@ object GcToolResults {
                 toolResultsStep.stepId
             )
             .setPageToken(pageToken)
+            .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, toolResultsStep.projectId))
             .executeWithRetry()
     }
 
@@ -155,12 +163,26 @@ object GcToolResults {
     }
 
     fun getDefaultBucket(projectId: String, source: String? = null): String? = try {
-        service.Projects().initializeSettings(projectId).executeWithRetry().defaultBucket
+        service.Projects().initializeSettings(projectId)
+            .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, projectId))
+            .executeWithRetry()
+            .defaultBucket
     } catch (ftlProjectError: FTLProjectError) {
         // flank needs to rewrap the exception with additional info about project
         when (ftlProjectError) {
-            is PermissionDenied -> throw FlankGeneralError(permissionDeniedErrorMessage(projectId, source, ftlProjectError.message))
-            is ProjectNotFound -> throw FlankGeneralError(projectNotFoundErrorMessage(projectId, ftlProjectError.message))
+            is PermissionDenied -> throw FlankGeneralError(
+                permissionDeniedErrorMessage(
+                    projectId,
+                    source,
+                    ftlProjectError.message
+                )
+            )
+            is ProjectNotFound -> throw FlankGeneralError(
+                projectNotFoundErrorMessage(
+                    projectId,
+                    ftlProjectError.message
+                )
+            )
             is FailureToken -> UserAuth.throwAuthenticationError()
         }
     }
@@ -190,6 +212,7 @@ object GcToolResults {
         )
         .setPageToken(pageToken)
         .setPageSize(100)
+        .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, results.projectId))
         .executeWithRetry()
 
     fun listAllSteps(results: ToolResultsExecution): MutableList<Step> {
@@ -217,6 +240,7 @@ object GcToolResults {
         )
         .setPageToken(pageToken)
         .setPageSize(100)
+        .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, results.projectId))
         .executeWithRetry()
 }
 

--- a/test_runner/src/main/kotlin/ftl/client/google/IosCatalog.kt
+++ b/test_runner/src/main/kotlin/ftl/client/google/IosCatalog.kt
@@ -1,9 +1,11 @@
 package ftl.client.google
 
+import com.google.api.client.http.HttpHeaders
 import com.google.testing.model.IosDeviceCatalog
 import com.google.testing.model.IosModel
 import com.google.testing.model.Orientation
 import ftl.config.Device
+import ftl.config.FtlConstants.GCS_PROJECT_HEADER
 import ftl.environment.getLocaleDescription
 import ftl.environment.ios.getDescription
 import ftl.environment.ios.iosVersionsToCliTable
@@ -50,6 +52,7 @@ object IosCatalog {
             GcTesting.get.testEnvironmentCatalog()
                 .get("ios")
                 .setProjectId(projectId)
+                .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, projectId))
                 .executeWithRetry()
                 .iosDeviceCatalog
                 .filterDevicesWithoutSupportedVersions()

--- a/test_runner/src/main/kotlin/ftl/client/google/OsVersion.kt
+++ b/test_runner/src/main/kotlin/ftl/client/google/OsVersion.kt
@@ -1,13 +1,16 @@
 package ftl.client.google
 
+import com.google.api.client.http.HttpHeaders
 import com.google.testing.model.AndroidVersion
 import com.google.testing.model.IosVersion
+import ftl.config.FtlConstants.GCS_PROJECT_HEADER
 import ftl.http.executeWithRetry
 
 fun androidOsVersions(projectId: String): List<AndroidVersion> =
     GcTesting.get.testEnvironmentCatalog()
         .get("android")
         .setProjectId(projectId)
+        .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, projectId))
         .executeWithRetry()
         .androidDeviceCatalog
         .versions
@@ -17,6 +20,7 @@ fun iosOsVersions(projectId: String): List<IosVersion> =
     GcTesting.get.testEnvironmentCatalog()
         .get("ios")
         .setProjectId(projectId)
+        .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, projectId))
         .executeWithRetry()
         .iosDeviceCatalog
         .versions

--- a/test_runner/src/main/kotlin/ftl/client/google/run/android/GcAndroidTestMatrix.kt
+++ b/test_runner/src/main/kotlin/ftl/client/google/run/android/GcAndroidTestMatrix.kt
@@ -1,5 +1,6 @@
 package ftl.client.google.run.android
 
+import com.google.api.client.http.HttpHeaders
 import com.google.common.annotations.VisibleForTesting
 import com.google.testing.Testing
 import com.google.testing.model.Account
@@ -22,6 +23,7 @@ import ftl.api.TestMatrixAndroid
 import ftl.client.google.GcTesting
 import ftl.client.google.run.toClientInfoDetailList
 import ftl.client.google.run.toFileReference
+import ftl.config.FtlConstants.GCS_PROJECT_HEADER
 import ftl.http.executeWithRetry
 import ftl.run.exception.FlankGeneralError
 import ftl.util.timeoutToSeconds
@@ -45,7 +47,9 @@ private suspend fun executeAndroidTestMatrix(
 ): List<Deferred<TestMatrix>> = coroutineScope {
     (0 until config.repeatTests).map { runIndex ->
         async(Dispatchers.IO) {
-            createAndroidTestMatrix(type, config, typeIndex, runIndex).executeWithRetry()
+            createAndroidTestMatrix(type, config, typeIndex, runIndex)
+                .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, config.project))
+                .executeWithRetry()
         }
     }
 }

--- a/test_runner/src/main/kotlin/ftl/client/google/run/ios/GcIosTestMatrix.kt
+++ b/test_runner/src/main/kotlin/ftl/client/google/run/ios/GcIosTestMatrix.kt
@@ -1,5 +1,6 @@
 package ftl.client.google.run.ios
 
+import com.google.api.client.http.HttpHeaders
 import com.google.testing.Testing
 import com.google.testing.model.ClientInfo
 import com.google.testing.model.EnvironmentMatrix
@@ -15,6 +16,7 @@ import ftl.client.google.run.mapGcsPathsToFileReference
 import ftl.client.google.run.mapToIosDeviceFiles
 import ftl.client.google.run.toClientInfoDetailList
 import ftl.client.google.run.toIosDeviceFile
+import ftl.config.FtlConstants.GCS_PROJECT_HEADER
 import ftl.http.executeWithRetry
 import ftl.run.exception.FlankGeneralError
 import ftl.util.timeoutToSeconds
@@ -37,7 +39,9 @@ private suspend fun executeIosTestMatrixAsync(
     config: TestMatrixIos.Config
 ): Deferred<TestMatrix> = coroutineScope {
     async(Dispatchers.IO) {
-        createIosTestMatrix(type, config).executeWithRetry()
+        createIosTestMatrix(type, config)
+            .setRequestHeaders(HttpHeaders().set(GCS_PROJECT_HEADER, config.project))
+            .executeWithRetry()
     }
 }
 

--- a/test_runner/src/main/kotlin/ftl/config/FtlConstants.kt
+++ b/test_runner/src/main/kotlin/ftl/config/FtlConstants.kt
@@ -25,6 +25,7 @@ object FtlConstants {
     const val indent = "  "
     const val matrixIdsFile = "matrix_ids.json"
     const val applicationName = "Flank"
+    const val GCS_PROJECT_HEADER = "x-goog-user-project"
     const val GCS_PREFIX = "gs://"
     const val GCS_STORAGE_LINK = "https://console.developers.google.com/storage/browser/"
     const val runTimeout = "-1"

--- a/test_runner/src/test/kotlin/ftl/reports/outcome/BillableMinutesTest.kt
+++ b/test_runner/src/test/kotlin/ftl/reports/outcome/BillableMinutesTest.kt
@@ -47,6 +47,7 @@ class BillableMinutesTest {
                 .testEnvironmentCatalog()
                 .get(any())
                 .setProjectId(any())
+                .setRequestHeaders(any())
                 .executeWithRetry()
         } returns make {
             androidDeviceCatalog = make { models = androidModels }

